### ISSUE TITLE
Add LLaMA support

### ISSUE THREAD

**Issue by ArthurConmy** [neelnanda-io/TransformerLens#201]
From Neel: we should add support for LLaMA and T5 (issue here: https://github.com/neelnanda-io/TransformerLens/issues/94 ).

LLaMA isn't open source so maybe with instructions to download from local weights.

(T5's architecture is a bit different (I think) but it's also somewhat useful as google have lots of models that are pretty powerful and use the same codebase, e.g https://huggingface.co/google/flan-ul2 )

**Comment by afspies** [comment]
How much work do we think this would be?

**Comment by ArthurConmy** [comment]
Isn't cross-attention in T5 only? If so, discussion should be moved here: https://github.com/neelnanda-io/TransformerLens/issues/94

**Comment by jmsdao** [comment]
A LLaMA implementation was merge into HF recently: https://github.com/huggingface/transformers/pull/21955/files#diff-06392bad3b9e97be9ade60d4ac46f73b6809388f4d507c2ba1384ab872711c51

**Comment by neelnanda-io** [comment]
Doesn't look like LLAMA has cross attention? You'll need to add RMSNorm and
RMSNormPre, and a GatedMLP block for SwiGLU. Shouldn't be too hard, rotary
is already implemented, though you might need to set rotary_dim=1., and I
don't know if it uses the GPTJ or NeoX striding (they pair up elements
differently...). The key thing will be having access to LLAMA weights and
code to run it, so you can check that your implementation works. Though you
could try taking Facebook's or HuggingFace's code, randomly initialising
weights, and verifying with those? It'll be easier if your model is smaller
than 7B for testing lol.

Tldr, not conceptually very hard, but will be tedious and involve a fair
bit of debugging and reading their code, since there's a lot of minor
variation in implementation details and you need to be EXACTLY right. You
want to take an existing implementation and compare exactly to their hidden
activations

**Comment by neelnanda-io** [comment]
Note that LLAMA is generative so can be added to the main HookedTransformer
Class . T5 will be much involved, and I think it should be a separate
class, which involves a lot more overhead
